virtual-host-gatherer/lib/gatherer/modules/Xen.py

```diff
@@ -0,0 +1,169 @@
+            self._validate_parameters(node)
+        except AttributeError as error:
+            self.log.error(error)
+            raise error
+
+        self.host = node['hostname']
+        self.port = node.get('port', 80)
+        self.user = node['username']
+        self.password = node['password']
+
+    def parameters(self):
+        """
+        Return default parameters
+
+        :return: default parameter dictionary
+        """
+
+        return self.DEFAULT_PARAMETERS
+
+    def run(self):
+        """
+        Start worker.
+
+        :return: Dictionary of the hosts in the worker scope.
+        """
+        output = dict()
+        self.log.info('Connect Xen Server %s:%s as user %s', self.host, self.port, self.user)
+        try:
+            session = XenAPI.Session("http://%s:%s" % (self.host, self.port))
+            session.xenapi.login_with_password(self.user, self.password, "2.3", "Xen Gatherer")
+
+            # Dictionary {opaque_reference: host_data}
+            hosts = {ref: session.xenapi.host.get_record(ref) for ref in session.xenapi.host.get_all()}
+            for _ref, host in hosts.items():
+                self.log.debug('Host=%s, uuid=%s', host["name_label"], host["uuid"])
+
+            vms = []
+            for vm_ref in session.xenapi.VM.get_all():
+                vm = session.xenapi.VM.get_record(vm_ref)
+                if not vm['is_a_template']:
+                    vms.append(vm)
+
+            for ref, host in hosts.items():
+                output[host['name_label']] = {
+                    'name': host['name_label'],
+                    'hostIdentifier': host['name_label'],
+                    'type': 'xen',
+                    'os': 'Xen AHV',
+                    'osVersion': host['software_version']['product_version'],
+                    'totalCpuSockets': host['cpu_info']['socket_count'],
+                    'totalCpuCores': host['cpu_info']['cpu_count'],
+                    'totalCpuThreads': 1,
+                    'cpuMhz': float(host['cpu_info']['speed']),
+                    'cpuDescription': host['cpu_info']['modelname'],
+                    'cpuArch': "x86_64",
+                    'ramMb': int(session.xenapi.host_metrics.get_record(host['metrics'])['memory_total']) / (1024 * 1024),
+                    'vms': {},
+                    'optionalVmData': {}
+                }
+
+                for vm in filter(lambda x: x['resident_on'] == ref, vms):
+                    output[host['name_label']]['vms'][vm['name_label']] = vm['uuid']
+                    output[host['name_label']]['optionalVmData'][vm['name_label']] = {}
+                    output[host['name_label']]['optionalVmData'][vm['name_label']]['vmState'] = self.VMSTATE.get(vm['power_state'], 'unknown')
+
+            output['Xen-DetachedVMs'] = {
+                'name': "Xen-DetachedVMs",
+                'hostIdentifier': "Xen-DetachedVMs",
+                'type': 'xen',
+                'os': 'Xen AHV',
```

Comment:
Does "Xen AHV" really exist?

AFAIK `AHV` stands for `Acropolis Hypervisor` which is the Nutanix product.

Reply:
I just checked on a Citrix XenServer Express installation I have here and this is what it reports:

You can see the base OS is a RHEL clone:

```
[root@xen ~]# ls -alh /etc/*release
-rw-r--r-- 1 root root  48 12 jun  2020 /etc/centos-release
-rw-r--r-- 1 root root 242 12 jun  2020 /etc/os-release
lrwxrwxrwx 1 root root  14 10 mar 23:13 /etc/redhat-release -> centos-release
lrwxrwxrwx 1 root root  14 10 mar 23:13 /etc/system-release -> centos-release
```

But it reports its name as "Citrix Hypervisor", and that's probably what we should report for the official Citrix Xen Hypervisor. If you want to go the extra mile, you can implement detecting whether it's Citrix or XCP-NG and reporting different 'os' but I don't think it's necessary:

```
[root@xen ~]# cat /etc/os-release
NAME="Citrix Hypervisor"
VERSION="8.2.0"
ID="xenenterprise"
ID_LIKE="centos rhel fedora"
VERSION_ID="8.2.0"
PRETTY_NAME="Citrix Hypervisor 8.2.0"
ANSI_COLOR="0;31"
HOME_URL="http://xenserver.org/"
BUG_REPORT_URL="https://bugs.xenserver.org/"
```

The `/etc/centos-release` is there but it's not CentOS:

```
[root@xen ~]# cat /etc/centos-release
Citrix Hypervisor release 8.2.0 (xenenterprise)
```

TL;DR: I would return `'os': 'citrixhypervisor'`, or `'os': 'xenhypervisor'` or alike.